File: src/leetcodepython/array/solution_view_40.py
# encoding='utf-8'

'''
面试题目40
author:zhangyu
date:2020.3.20
'''
from typing import List

from queue import PriorityQueue as PQueue


class Solution:
    def get_least_numbers(self, arr: List[int], k: int) -> List[int]:
        '''
            获取最小k个数
        Args:
            arr: 数组
            k: 数个数
        Returns:
            最小k个数
        '''
        arr.sort()
        return arr[:k]

    def get_least_numbers2(self, arr: List[int], k: int) -> List[int]:
        '''
            获取最小k个数
        Args:
            arr: 数组
            k: 数个数
        Returns:
            最小k个数
        '''
        pq = PQueue()
        new_arr = []
        for num in arr:
            pq.put(num)
            if pq.qsize() > len(arr) - k:
                new_arr.append(pq.get())
        return new_arr

    def swap(self, arr: List[int], i: int, j: int) -> None:
        '''
            交换数组中两个元素的位置
        Args:
            arr: 数组
            i: 位置i
            j: 位置j
        Returns:

        '''
        temp = arr[i]
        arr[i] = arr[j]
        arr[j] = temp

    def partition(self, arr: List[int], left: int, right: int) -> int:
        '''
            找到分割点位置
        Args:
            arr: 数组
            left: 左边
            right: 右边
        Returns:
            下标位置
        '''
        pivot, index = arr[left], left
        for i in range(left + 1, right + 1):
            if arr[i] < pivot:
                index += 1
                self.swap(arr, index, i)
        self.swap(arr, left, index)
        return index

    def get_least_numbers3(self, arr: List[int], k: int) -> List[int]:
        '''
            获取最小k个数
        Args:
            arr: 数组
            k: 数个数
        Returns:
            最小k个数
        '''
        length = len(arr)
        if k == 0 or k > length:
            return []
        target, left, right = k - 1, 0, length - 1
        while True:
            index = self.partition(arr, left, right)
            if index == target:
                return arr[:k]
            elif index < target:
                left = index + 1
            else:
                right = index - 1


if __name__ == '__main__':
    arr = [3, 4, 5, 6, 1, 3]
    k = 2
    solution = Solution()
    new_arr = solution.get_least_numbers3(arr, k)
    print(new_arr)

File: src/leetcodejava/array/SolutionView40.java
package leetcodejava.array;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

/**
 * 面试40:最小的k个数
 *
 * @author: zhangyu
 */
public class SolutionView40 {

    @Test
    public void testSolutionView() {
        int[] arr = {3, 4, 5, 6, 1, 3, 5, 6, 7, 9};
        int k = 3;
        int[] partition = getLeastNumbers3(arr, k);
        System.out.println(partition);
    }

    /**
     * 获取最小多个值;快速排序
     *
     * @param arr 数组
     * @param k   k个数
     * @return 数组
     */
    public int[] getLeastNumbers(int[] arr, int k) {
        if (arr.length == 0) {
            return arr;
        }
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length - 1 - i; j++) {
                if (arr[j + 1] < arr[j]) {
                    int temp = arr[j + 1];
                    arr[j + 1] = arr[j];
                    arr[j] = temp;
                }
            }
        }
        return Arrays.stream(arr).limit(k).toArray();
    }

    /**
     * 获取最小多个值,利用优先队列
     *
     * @param arr 数组
     * @param k   k个数
     * @return 数组
     */
    public int[] getLeastNumbers2(int[] arr, int k) {
        PriorityQueue<Integer> priorityQueue = new PriorityQueue();
        List<Integer> list = new ArrayList<>();
        for (int num : arr) {
            priorityQueue.add(num);
            if (priorityQueue.size() > arr.length - k) {
                list.add(priorityQueue.poll());
            }
        }
        return list.stream().mapToInt(e -> Integer.valueOf(e)).toArray();
    }

    /**
     * 获取最小多个值 ,利用快速排序方法
     *
     * @param arr 数组
     * @param k   k个数
     * @return 数组
     */
    public int[] getLeastNumbers3(int[] arr, int k) {
        int len = arr.length;
        if (k == 0 || k > len) {
            return new int[0];
        }
        int target = k - 1;
        int left = 0;
        int right = len - 1;
        while (true) {
            int index = partition(arr, left, right);
            if (index == target) {
                int[] res = new int[k];
                for (int i = 0; i < k; i++) {
                    res[i] = arr[i];
                }
                return res;
            } else if (index < target) {
                // 下一轮搜索区间在 [index + 1, right]
                left = index + 1;
            } else {
                // 下一轮搜索区间在 [left, index - 1]
                right = index - 1;
            }
        }
    }

    /**
     * 进行分区操作,是的左边小于阈值，右边大于等于阈值
     *
     * @param arr   数组
     * @param left  左边
     * @param right 右边
     * @return 下标
     */
    private int partition(int[] arr, int left, int right) {
        int pivot = arr[left];
        int index = left;
        for (int i = left + 1; i <= right; i++) {
            if (arr[i] < pivot) {
                index++;
                swap(arr, index, i);
            }
        }
        swap(arr, left, index);
        return index;
    }

    /**
     * 交换数据中元素的位置
     *
     * @param arr 数组
     * @param i   位置i
     * @param j   位置j
     */
    private void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
